### DISCUSSION TITLE
java: package javafx.application does not exist

### DISCUSSION THREAD

**Discussion by PerryCameron** [wiverson/maven-jpackage-template#79] [Q&A]
One thing I have found strange with this setup is the inability to use a run configuration from the IDE. I am sure I am missing something, but I am able to use "mvn javafx:run" in the terminal of intellij and the program will launch. If however, I try to set up a run configuration and run it directly from the IDE I get a bunch of "java: package javafx.application does not exist" style errors. I find this strange since I am using 19.fx-zulu where the packages are included. It is acting like it is looking for modules. Do you have any clues to what may be going on here?

**Comment by wiverson** [comment]
Yup.

To fix, you will need to turn off download shared indexes in IntelliJ and invalidate caches.

Preferences -> Tools -> Public Shared Indexes -> JDKs -> Don't Download
File -> Invalidate Caches or File -> Repair IDE

tl;dr - the shared index IntelliJ downloads does not include JavaFX.

FWIW Once that's all done you should be able to run & debug just by launching the main() method, e.g. BaseApplication.main. You can click the run button in the gutter the first time and then run/debug after that.

**Reply by PerryCameron** [reply]
I changed those to both don't download. Then I want to Invalidate Caches and checked all the boxes. The IDE reset and a popup came up asking me if I wanted to download the indexes. I clicked on "more" and don't show this again.  I then noticed in the status bar that it seemed to be indexing 19.fx-zulu. When I went to the main method and launched...exact same thing, it doesn't see those libraries. I don't mind using mvn javafx:run but I need to be able to see exceptions. Any more ideas?

**Reply by wiverson** [reply]
Make sure the Project is pointing at the right JDK. File -> Project Structure... -> Project Settings -> Project -> SDK. Make sure it is pointing at the correct SDK.

You might need to make sure the File -> Project Structure... -> Platform Settings -> SDKs includes the right JDK as well.

**Reply by PerryCameron** [reply]
There is no issue there the SDK is set fine. Looking through YouTrack from 3 years ago I found this relevant link.
https://youtrack.jetbrains.com/issue/IDEA-224706/IntelliJ-unable-to-compile-code-which-uses-javafx-when-I-use-LibericaJDK-13

Sergei Tachenov (bottom comment) said: 
It seems to me that this is not really a bug, but rather a confusing feature. The --release option, as far as I understand it, checks the used API against the specified API version (not a particular implementation). That is, with this option, the sources are only allowed to access explicit dependencies (from pom.xml, for example) and the standard part of the API. And JavaFX is not a part of the standard API. This is kind of makes sense, as if you happily compile an app with Full Liberica JDK without specifying explicit JavaFX dependency and then try to run it on some random machine under, say, AdoptJDK, you're going to have a problems. With the --release option you pretty much guarantee that it will run fine on any standard-compliant JRE. The only problem I see is that no errors are displayed in the editor. Having an import highlighted in red with a message along the lines of “This class is not available because the --release option prevents its usage” would certainly be helpful.

Now this issue would not matter to us because we are including the JDK to make the program executable. I am going to keep looking around, but If I can't find a way to see exceptions related to my code this is kind of a show stopper. Hopefully there is some sort of work around.

**Reply by wiverson** [reply]
Huh. Not sure what to say. I'm using Liberica+JavaFX+IntelliJ w/no issues after going through the above, and helped folks get it working on other machines following similar steps. I suppose you could try Liberica. Are you setting --release somewhere...?

**Reply by PerryCameron** [reply]
No, I haven't set that. Not even sure where it is.  I set it to 19.fx-librca and getting the same problem.

**Reply by PerryCameron** [reply]
Well, It launches for eclipse without any issue

**Reply by wiverson** [reply]
Huh. Not sure what going on. I would assume that it's something to do with your IntelliJ config but not sure what. :/

FWIW I left [this branch](https://github.com/wiverson/maven-jpackage-template/tree/jdk-17-custom-javafx) up as the last version that broke apart the JDK from the JavaFX jmods. You will have to be careful to match JavaFX versions w/ the JDK you are using, but it did seem to play nicer w/IntelliJ.

**Reply by PerryCameron** [reply]
I went ahead and put in the question to Intellij. We shall see what they come up with.

https://intellij-support.jetbrains.com/hc/en-us/community/posts/7945591296274-javafx-application-does-not-exist

**Reply by PerryCameron** [reply]
Well I moved it to Java 17 and now it works.
